File: Week 2/4LeastCommonMultiple.cpp
#include <iostream>

long long lcm_naive(int a, int b) {
	for (long l = 1; l <= (long long) a * b; ++l)
		if (l % a == 0 && l % b == 0)
			return l;

	return (long long) a * b;
}

long long lcm_smart(long long a, long long b) {
	long long a2 = a,
		b2 = b,
		product = 0,
		gcd,
		lcm;

	do {
		product = a2 % b2;
		a2 = b2;
		b2 = product;
	} while (product > 0);

	gcd = a2;
	lcm = a * b / gcd;

	return lcm;
}

int main() {
	long long a, b, temp;

	std::cin >> a >> b;

	// swap so a is highest
	if (b > a) {
		temp = a;
		a = b;
		b = temp;
	}

	std::cout << lcm_smart(a, b) << std::endl;
	return 0;
}

File: Week 2/4LeastCommonMultiple.js
var readline = require('readline');

process.stdin.setEncoding('utf8');
var rl = readline.createInterface({
  input: process.stdin,
  terminal: false
});

rl.on('line', readLine);

let a, b;
let product;

function readLine (line) {
	if (line !== "\n") {
		let ints = line.toString().split(' ').map(n => +n);
		if (ints[1] > ints[0]) {
			a = ints[1];
			b = ints[0];
		} else {
			a = ints[0];
			b = ints[1];
		}
	}
}

rl.on('close', () => {
	let newA = a;
	let newB = b;

	do {
		product = newA % newB;
		newA = newB;
		newB = product;
	} while (product > 0);

	// newB is the gcd now.
	let gcd = newA;

	let brandNewA = a / gcd;
	let brandNewB = b / gcd;

	// to avoid overflow
	let lcm = a * b / gcd;
	console.log(lcm);
});